### PR TITLE
Moi interface

### PR DESCRIPTION
**Pull request by vitornesello**
Able to define build, solve and get objective value. Supporting ScalarAffineFunction and SingleVariable.   All tests are passing.

### CODE REVIEW

File: src/wrapperfunctions/MOIWrapper.jl [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+function add_memberships(constr::Constraint, coluna_vars::Vector{MasterVar},
+                         problem::Problem, f::MOI.ScalarAffineFunction,
+                         mapping::MOIU.IndexMap)
+    for term in f.terms
+        add_membership(coluna_vars[mapping.varmap[term.variable_index].value],
+                       constr, problem, term.coefficient)
+    end
+end
+
+function get_problem_from_constraint(dest::ColunaModelOptimizer, ci::MOI.ConstraintIndex)
```
Comment (IssamT) [inline src/wrapperfunctions/MOIWrapper.jl]: Maybe better name it get_constraint_problem.

File: src/wrapperfunctions/MOIWrapper.jl [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    name = string("constraint_", mapping.conmap[ci].value)
+    constr = MasterConstr(problem.counter, name,
+                          s.lower - f.constant, 'G', 'M', 's')
+
+    add_constraint(problem, constr)
+    add_memberships(constr, coluna_vars, problem, f, mapping)
+
+end
+
+function load_constraint(ci::MOI.ConstraintIndex, dest::ColunaModelOptimizer,
```
Comment (IssamT) [inline src/wrapperfunctions/MOIWrapper.jl]: load_constraint methods could use a helper function to reduce code redundacy

File: src/wrapperfunctions/MOIWrapper.jl [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+            for ci in MOI.get(src, MOI.ListOfConstraintIndices{F,S}())
+                f = MOI.get(src, MOI.ConstraintFunction(), ci)
+                s = MOI.get(src,  MOI.ConstraintSet(), ci)
+                load_constraint(ci, dest, mapping, coluna_vars, f, s)
+            end
+        end
+    end
+end
+
+function copy_singleVariable_constraints(dest::ColunaModelOptimizer, src::MOI.ModelLike,
```
Comment (IssamT) [inline src/wrapperfunctions/MOIWrapper.jl]: don't use capital letters in name

File: src/wrapperfunctions/MOIWrapper.jl [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+                         f::MOI.SingleVariable, s::MOI.Interval)
+    if s.lower > coluna_vars[mapping.varmap[f.variable].value].lower_bound
+        coluna_vars[mapping.varmap[f.variable].value].lower_bound = s.lower
+    end
+    if s.upper < coluna_vars[mapping.varmap[f.variable].value].upper_bound
+        coluna_vars[mapping.varmap[f.variable].value].upper_bound = s.upper
+    end
+end
+
+function copy_constraints(dest::ColunaModelOptimizer, src::MOI.ModelLike,
```
Comment (IssamT) [inline src/wrapperfunctions/MOIWrapper.jl]: this is not generic. Add something in the name to show the difference with the next one

File: src/wrapperfunctions/MOIWrapper.jl [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+        problem = dest.inner.extended_problem.master_problem
+        add_variable(problem, coluna_vars[idx])
+    end
+end
+
+function build_constraint_mapping(mapping::MOIU.IndexMap, src::MOI.ModelLike)
+    list_of_constraints = MOI.get(src, MOI.ListOfConstraints())
+    num_rows = 0
+    for (F,S) in list_of_constraints
+        # if !(MOI.supportsconstraint(dest, F, S))
```
Comment (IssamT) [inline src/wrapperfunctions/MOIWrapper.jl]: Please delete this code if it is not needed.

File: src/wrapperfunctions/MOIWrapper.jl [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+    function copy!(coluna_optimizer, user_optimizer; copynames=false)
+
+    ## Create variables without adding to problem
+    ## Update the variable cost_rhs
+    ## Go through SingleVariable constraints and modify the variables
+    ## Add variables to problem
+    ## Go throught ScalarAffineFunction constraints
+
+"""
+function MOI.copy!(dest::ColunaModelOptimizer, src::MOI.ModelLike; copynames=false)
```
Comment (IssamT) [inline src/wrapperfunctions/MOIWrapper.jl]: We don't add the doc string when overriding MOI functions

File: test/test_utils.jl [hunk 7]
```diff
@@ -1,50 +1,118 @@
 @MOIU.model ModelForCachingOptimizer (ZeroOne, Integer) (EqualTo, GreaterThan, LessThan, Interval) () () (SingleVariable,) (ScalarAffineFunction,) () ()
 
+
+function build_tricky_model()
+
+    coluna_optimizer = CL.ColunaModelOptimizer()
+
+    moi_model = MOIU.CachingOptimizer(ModelForCachingOptimizer{Float64}(),
+                                      coluna_optimizer)
+
```
Comment (IssamT) [inline test/test_utils.jl]: Don't put two consecutive empty lines in the code.

File: src/wrapperfunctions/MOIWrapper.jl [hunk 8]
```diff
@@ -0,0 +1,398 @@
+export ColunaModelOptimizer
+
+
+mutable struct ColunaModelOptimizer <: MOI.AbstractOptimizer
+    inner::Model
+    function ColunaModelOptimizer()
+        coluna_model = ModelConstructor()
```
Comment (IssamT) [inline src/wrapperfunctions/MOIWrapper.jl]: Where is ModelConstructor defined?

Comment (vitornesello) [reply src/wrapperfunctions/MOIWrapper.jl]: Defined in Coluna. In file model.jl:14

File: src/wrapperfunctions/MOIWrapper.jl (hunk 5 above)
Comment (vitornesello) [reply src/wrapperfunctions/MOIWrapper.jl]: I will use it later when I implement the supportsconstraint functions. Can I leave it there?

File: src/wrapperfunctions/MOIWrapper.jl (hunk 1 above)
Comment (vitornesello) [reply src/wrapperfunctions/MOIWrapper.jl]: This is better.

File: src/wrapperfunctions/MOIWrapper.jl (hunk 3 above)
Comment (vitornesello) [reply src/wrapperfunctions/MOIWrapper.jl]: Done.

File: src/wrapperfunctions/MOIWrapper.jl (hunk 2 above)
Comment (vitornesello) [reply src/wrapperfunctions/MOIWrapper.jl]: Done.

File: src/wrapperfunctions/MOIWrapper.jl (hunk 4 above)
Comment (vitornesello) [reply src/wrapperfunctions/MOIWrapper.jl]: Done.

File: src/wrapperfunctions/MOIWrapper.jl [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+end
+
+# function MOI.get(coluna_optimizer::ColunaModelOptimizer, object::MOI.NodeCount)
+# end
+
+function MOI.get(coluna_optimizer::ColunaModelOptimizer, object::MOI.ObjectiveValue)
+    return coluna_optimizer.inner.extended_problem.primal_inc_bound
+end
+
+# function MOI.get(coluna_optimizer::ColunaModelOptimizer, object::MOI.VariablePrimal, ref::MOI.VariableIndex)
```
Comment (IssamT) [inline src/wrapperfunctions/MOIWrapper.jl]: These commented methods can be always copied later. No?

**Comment by IssamT** [comment]
Great job Vitor! I like the interface and the tests are clear :)

File: src/MOIWrapper.jl [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+    end
+end
+
+## Annotations needed for column generation
+## Problem Index: if 0 -> Master problem, if > 0 -> subproblem index
+struct ConstraintDantzigWolfeAnnotation <: MOI.AbstractConstraintAttribute end
+
+function MOI.set!(dest::MOIU.UniversalFallback, attribute::ConstraintDantzigWolfeAnnotation,
+                  ci::MOI.ConstraintIndex, value::Int)
+    if haskey(dest.conattr, attribute)
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/MOIWrapper.jl]: This is from UniversalFallback

File: src/MOIWrapper.jl [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+end
+
+function MOI.get(dest::MOIU.UniversalFallback, attribute::ConstraintDantzigWolfeAnnotation,
+                 ci::MOI.ConstraintIndex)
+    if haskey(dest.conattr, attribute)
+        if haskey(dest.conattr[attribute], ci)
+            return dest.conattr[attribute][ci]
+        end
+    end
+    return -1 ## Returns value -1 as default if not found
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/MOIWrapper.jl]: What if the user is solving only branch and bound? Will he need to define the annotations anyway?

File: src/MOIWrapper.jl [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+function MOI.optimize!(coluna_optimizer::ColunaModelOptimizer)
+    solve(coluna_optimizer.inner)
+end
+
+function load_obj(dest::ColunaModelOptimizer, coluna_vars::Vector{<:Variable},
+                  mapping::MOIU.IndexMap, f::MOI.ScalarAffineFunction)
+    # We need to increment values of cost_rhs with += to handle cases like $x_1 + x_2 + x_1$
+    # This is safe becasue the variables are initialized with a 0.0 cost_rhs
+    for term in f.terms
+        coluna_vars[mapping.varmap[term.variable_index].value].cost_rhs += term.coefficient
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/MOIWrapper.jl]: Better use dest.varmap. I removed from everywhere that is not really usefull

File: src/MOIWrapper.jl [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+end
+
+function copy_constraints(dest::ColunaModelOptimizer, src::MOI.ModelLike,
+    mapping::MOIU.IndexMap, coluna_vars::Vector{<:Variable}; only_singlevariable = false)
+    for (F,S) in MOI.get(src, MOI.ListOfConstraints())
+        if (F == MOI.SingleVariable && only_singlevariable
+            || F != MOI.SingleVariable && !only_singlevariable)
+            for ci in MOI.get(src, MOI.ListOfConstraintIndices{F,S}())
+                f = MOI.get(src, MOI.ConstraintFunction(), ci)
+                s = MOI.get(src,  MOI.ConstraintSet(), ci)
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/MOIWrapper.jl]: where is the extra space exactly?

File: src/MOIWrapper.jl [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+    if sense != MOI.MinSense
+        error("Minimization is the only supported sense for now.")
+    end
+end
+
+function MOI.empty!(coluna_optimizer::ColunaModelOptimizer)
+    coluna_optimizer.inner = ModelConstructor()
+end
+
+# ## Get functions
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/MOIWrapper.jl]: Ok, I am fine with your suggestion

[the remark this answers is not in the document]
Comment (vitornesello) [reply src/MOIWrapper.jl]: But sometimes we want to give some more emphasis on a comment, for example when we want to write few lines of comments. Should we use only one even then?

File: src/TAGS [hunk 15]
```diff
@@ -0,0 +1,240 @@
+
+constraints.jl,117
+function MasterConstrBuilder(25,895
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/TAGS]: soryy, I commited this file by mistake. I will add to gitignore

File: src/MOIWrapper.jl [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+    dest.nb_subproblems = find_number_of_subproblems(src)
+    for i in 1:dest.nb_subproblems
+        subprob = SimpleCompactProblem(prob_counter, counter)
+        push!(extended_problem.pricing_vect, subprob)
+    end
+end
+
+function MOI.copy!(dest::ColunaModelOptimizer, src::MOI.ModelLike; copynames=false)
+    if copynames
+        error("Copynames not supported yet")
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/MOIWrapper.jl]: But copynames is not an attribute or anything. I will implement the feature of copynames and thats it

File: src/MOIWrapper.jl [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+            prob_idx = MOI.get(src, ConstraintDantzigWolfeAnnotation(), ci)
+            if prob_idx > nb_subproblems
+                nb_subproblems = prob_idx
+            end
+        end
+    end
+    return nb_subproblems
+end
+
+function create_subproblems(dest::ColunaModelOptimizer, src::MOI.ModelLike)
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/MOIWrapper.jl]: I don't think we need to check because they are generated contiguously with `push!`

**Comment by IssamT** [comment]
Also, the CI needs to be green ;)

**Pull request merged**